src/range/range.rs

```diff
@@ -0,0 +1,162 @@
+use crate::headers::{HeaderName, HeaderValue, Headers, ToHeaderValues, RANGE};
+use crate::range::{bytes, BytesRangeSet};
+use crate::{Error, StatusCode};
+
+use std::fmt::{self, Debug, Display};
+use std::option;
+
+/// HTTP Range request header.
+///
+/// Range header in a GET request modifies the method
+/// semantics to request transfer of only one or more subranges of the
+/// selected representation data, rather than the entire selected
+/// representation data.
+///
+/// # Specifications
+///
+/// - [RFC 7233, section 3.1: Range](https://tools.ietf.org/html/rfc7233#section-3.1)
+/// - [RFC 7233, Appendix D: Collected ABNF](https://tools.ietf.org/html/rfc7233#appendix-D)
+/// - [IANA HTTP parameters, range-units: HTTP Range Unit Registry](https://www.iana.org/assignments/http-parameters/http-parameters.xhtml#range-units)
+///
+/// # Examples
+///
+/// Range header using bytes:
+///
+/// ```
+/// # fn main() -> http_types::Result<()> {
+/// #
+/// use http_types::range::{BytesRangeSet, Range};
+/// use http_types::{Method, Request, Url};
+///
+/// let mut byte_range_set = BytesRangeSet::new();
+/// byte_range_set.push(0, 500);
+///
+/// let range = Range::Bytes(byte_range_set.clone());
+/// let mut req = Request::new(Method::Get, Url::parse("http://example.com").unwrap());
+/// range.apply(&mut req);
+///
+/// let range = Range::from_headers(req)?.unwrap();
+/// assert_eq!(range, Range::Bytes(byte_range_set));
+/// #
+/// # Ok(()) }
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum Range {
+    /// Bytes based range requests.
+    Bytes(BytesRangeSet),
+}
+
+impl Range {
+    /// Create a new instance from a Range headers.
+    ///
+    /// Only a single Range per resource is assumed to exist. If multiple Range
+    /// headers are found the last one is used.
+    pub fn from_headers(headers: impl AsRef<Headers>) -> crate::Result<Option<Self>> {
+        let headers = match headers.as_ref().get(RANGE) {
+            Some(headers) => headers,
+            None => return Ok(None),
+        };
+
+        // If a header is returned we can assume at least one exists.
+        let s = headers.iter().last().unwrap().as_str();
+
+        Self::from_str(s).map(Some)
+    }
+
+    /// Create a ByteRanges from a string.
+    pub(crate) fn from_str(s: &str) -> crate::Result<Self> {
+        let fn_err = || {
+            Error::from_str(
+                StatusCode::BadRequest,
+                "Invalid Range header for byte ranges",
```

Comment:
In my previous commits I add several error cases to handle, let closure here is no more justified I should move this block to the match block below